Baw-Client Migration

Allcharles opened the pull request:
# Migration of the Baw Client

This is the initial work on implementing the migration of the baw client into the workbench-client. The approach taken was to use iframes, running the baw-client from the assets folder of the workbench-client.

## Changes

- ~Modified app and home components to enable full size page components~
- Created audio analysis pages (mostly a placeholder for testing)
  - `ListComponent`
  - `NewComponent`
- ~Created iframe component (rename to `baw-client`?)~
  - Created `baw-client` component
- Inserted build assets from baw-client into `./src/assets/old-client`
  - Use the [`migration`](https://github.com/QutEcoacoustics/baw-client/tree/migration) branch of baw-client
  - Run `npm run build -- --development`
  - Copy build files inside `./bin` folder to `./src/assets/old-client/`

## Problems

- Some baw-client pages do not resize correctly. This may be improved by modifying more global styles, however it is likely that the styles for individual pages may need to be edited (example new audio analysis page having a horizontal scroll).
- Some visual design problems in the audio analysis list and new pages (these are intentionally left for a separate PR as they are only for testing)
- Need to develop a standardized method of testing components which will use the `IframeComponent`. This will be left for a separate PR
- [x] Tests will not pass until there is a method of mocking the baw-client build artifacts on the VM
- [x] Some tests still need to be implemented such as for the baw interceptor

## Visual Changes

### Audio Analysis Page

![image](https://user-images.githubusercontent.com/3955116/103500824-fad19700-4e97-11eb-87d5-c539c9fc9fda.png)

### New Audio Analysis Page

![image](https://user-images.githubusercontent.com/3955116/103500920-5865e380-4e98-11eb-9d73-c168664f4dff.png)


## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing
- [x] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Migration of the Baw Client

This is the initial work on implementing the migration of the baw client into the workbench-client. The approach taken was to use iframes, running the baw-client from the assets folder of the workbench-client.

Allcharles committed:
Build Baw Client (#839)

Build baw client in github actions, this allows the unit tests to validate that the baw client has been correctly integrated into the project.

Allcharles committed:
Migrate Baw Client Routes (#780)

Create all of the routing required to handle displaying the baw client pages. This makes use of the `BawClientComponent` created in #727.

Allcharles commented on src/app/app.component.scss:
```diff
@@ -1,32 +1,92 @@
+      }
+    }
+  }
 }
```
This needs commenting to explain what is doing what

Allcharles commented on src/app/app.component.ts:
```diff
@@ -15,15 +14,16 @@ import { AppConfigService } from "./services/app-config/app-config.service";
   styleUrls: ["./app.component.scss"],
+  // Need this so that router-outlet components can be styled
+  // eslint-disable-next-line @angular-eslint/use-component-view-encapsulation
+  encapsulation: ViewEncapsulation.None,
```
Not sure if this is need anymore. Reverted the changes which affected the sizing of page components

Allcharles commented on src/assets/environment.json:
```diff
@@ -2,7 +2,7 @@
   "environment": {
     "environment": "development",
     "apiRoot": "https://staging.ecosounds.org",
-    "siteRoot": "http://localhost:4200",
```
Revert this change? Running this default environment file only works with `npm run start:dev`

atruskie replied:
no this is a fine change - i've already suggested dev should bind to everything

maybe add a note to the readme that alerts the user that the dev server binds to all interfaces

atruskie commented on package.json:
```diff
@@ -13,6 +13,7 @@
     "ng": "node --max_old_space_size=4096 ./node_modules/@angular/cli/bin/ng",
     "nx": "nx",
     "start": "nx serve",
+    "start:dev": "nx serve --host 0.0.0.0 --disableHostCheck",
```
why not just change start to do what `start:dev` does?

Allcharles replied:
Mostly because its a potential security vulnerability for the running computer:
```
Warning: This is a simple server for use in testing or debugging Angular applications
locally. It hasn't been reviewed for security issues.

Binding this server to an open connection can result in compromising your application or
computer. Using a different host than the one passed to the "--host" flag might result in
websocket connection issues. You might need to use "--disableHostCheck" if that's the
case.
Warning: Running a server with --disable-host-check is a security risk. See https://medium.com/webpack/webpack-dev-server-middleware-security-issues-1489d950874a for more information.
```

atruskie replied:
Yeah, not really. The server is not meant to run in production. That's what they're discouraging 

Basically, if you're running any firewall it is totally fine. Even if not, then, it is likely fine.

Anything running from inside docker has to bind to any address as well because the containers are on different networks from the host and you wouldn't be able to use your dev service otherwise.

atruskie commented on src/app/app.component.html:
```diff
@@ -1,42 +1,23 @@
-          <router-outlet name="secondary"></router-outlet>
-        </div>
+  <div [ngClass]="{ container: !fullscreen, 'container-fluid': fullscreen }">
+    <!-- Secondary Menu (only active on menuLayout) -->
```
"only active on menuLayout" - what does this mean?

atruskie replied:
I'm guessing it means this outlet is used when the menu component is rendered? But given our routing setup, this happens the same time as the main router-outlet rendering, right?

Allcharles replied:
This is a mistake. A remnant of previous behavior where visibility was triggered by the action css class

atruskie commented on src/app/app.component.html:
```diff
@@ -1,42 +1,23 @@
-          <router-outlet></router-outlet>
-          <router-outlet name="error"></router-outlet>
-        </div>
+    <!-- Default and Error Menus -->
```
Is an error really a menu?

atruskie commented on src/app/app.component.scss:
```diff
@@ -1,32 +1,114 @@
-}
+  // Menu Layout
+  > .container {
+    // Compact menus on large displays
```
? wouldn't we want compact menu on small displays?

atruskie replied:
If I understand the `@include` media plugin correctly, it is doing that, just comment is misleading

atruskie commented on src/app/app.component.scss:
```diff
@@ -1,32 +1,114 @@
 
-  &:last-child {
-    min-height: 100%;
+    // Hide secondary/action menus when fullscreen
+    > #secondary,> #action {
+      display: none;
+    }
```
Confim, when hidden, menus are still accessible, e.g. via clicking a hamburger (menu) icon

Allcharles replied:
When hidden, menus are not accessible. Turning them into a hamburger menu is a future upgrade to the design of the website. This is tracked by: #458

atruskie commented on src/app/app.component.ts:
```diff
@@ -40,54 +44,13 @@ export class AppComponent extends withUnsubscribe() implements OnInit {
-    } else {
-      this.fullscreen = !!component.pageInfo.fullscreen;
-    }
+  public updatePageLayout(component: any) {
+    this.fullscreen = !!(component?.constructor?.pageInfo as PageInfo)
+      ?.fullscreen;
```
Since the mechanics of this mostly happen in templates, add a comment explaining workflow...

When component is loaded, uses its knowledge of pageInfo, to call this method (to escape encapsualtion) and change the fullscreen property, which is bound ??by/from/to what??

atruskie commented on src/app/components/audio-analysis/pages/list/list.component.spec.ts:
```diff
@@ -0,0 +1,23 @@
+  it("should create", () => {
+    expect(component).toBeTruthy();
+  });
+});
```
pointless test?

if it's just testing costruction, at least make the `toBeTruthy` expectation more useful. Why not test that component is an instance of AudioAnalysis.ListComponent ?

atruskie replied:
Another good test for these pages is asserting the `baw-client` element/component is exists/rendered in the template

Allcharles replied:
This test is disabled, (`xdescribe`). It is implemented later on, this is just the default test generated by angular schematics

atruskie commented on src/app/components/audio-analysis/pages/new/new.component.spec.ts:
```diff
@@ -0,0 +1,23 @@
+  });
+
+  it("should create", () => {
+    expect(component).toBeTruthy();
```
same comments on tests as with list.component.spec.ts

atruskie commented on src/app/components/shared/baw-client/baw-client.component.ts:
```diff
@@ -0,0 +1,81 @@
+})
+export class BawClientComponent extends withUnsubscribe() implements OnInit {
+  @ViewChild("content") private iframeRef: ElementRef<HTMLIFrameElement>;
+  public url: SafeResourceUrl;
```
this should probably be an`@input` so we can customise what baw-client page is loaded

atruskie reviewed (approved):
There's no details page for the audio-analyses?

> Some baw-client pages do not resize correctly. This may be improved by modifying more global styles, however it is likely that the styles for individual pages may need to be edited (example new audio analysis page having a horizontal scroll).

Ignore. Immediate priority is deploying, and then we'll be rebuilding old pages

> Some visual design problems in the audio analysis list and new pages (these are intentionally left for a separate PR as they are only for testing)

Probably ignore. Depends what the issue is.

> Need to develop a standardized method of testing components which will use the IframeComponent. This will be left for a separate PR

The only tests I want to see are ones that verify:

- the component loads
- the iframe loads
- there's non-error content in the iframe
- 
But little else

Allcharles committed:
Working Login (#853)

* Cleanup app styling to create full height components

- Cleanup CSS styling for the app component so that the default router outlet component will have access to the full height of the page
- Created basic audio analysis page

* Working example

Couple bugs to fix
- AngularJS links are not reloading the parent page
- IFrame does not resize to content which causes the scrollbar to be in the wrong location

* Created new audio analysis page

* Cleanup iframe module

- Disabled broken tests
- Renamed IFrameComponent to BawClientComponent
- Isolated BawClientComponent into a module
- Created placeholder unit tests
- Updated interceptor to include cookies in requests

* Implemented BawClientComponent unit tests

* Fixed broken tests and rebase conflicts

* Fixed broken unit tests

Discovered bug in how secondary menu unit tests were saving/restoring the defaultMenu global variable (converted Lists to arrays, but not back)

* Fix site/region details tag links

* Working login request

* Update security service tests

* Cleanup interceptor tests

* Fix rebase issues

* Cleanup api request responseType's

This should make things a lot more clear for when the interceptor should set accept/content-type headers to requests

* Cleanup review comments

Allcharles committed:
Reduce possibility of race condition failing unit tests
